### DISCUSSION TITLE
Setting PS1 variable

### DISCUSSION THREAD

**Discussion by jcameron** [webmin/webmin#1781] [Ideas]
https://github.com/webmin/webmin/blob/661bb957f9be57694b15f7b3c8491c3d16528f3b/xterm/index.cgi#L209

@iliajie why don't we just set this environment variable in shellserver.pl before running bash, rather than passing it through the terminal?

**Comment by swelljoe** [comment]
Why are we messing with people's environment at all? Shouldn't we respect people's .bashrc/.profile or whatever and not change colors and such? (I think I'm reading that we're adding our own colors to the shell, anyway, which we really shouldn't be doing. We respect user's choices.)

**Reply by iliaross** [reply]
> @iliajie why don't we just set this environment variable in shellserver.pl before running bash, rather than passing it through the terminal?

Jamie, nice suggestion. Fixed here https://github.com/webmin/webmin/commit/9cef9452d18f0b0dfb85effa2cf48bd9e3466326, and cleaned a bit later here https://github.com/webmin/webmin/commit/29517915f923d15fb36dd3aeecc973de016bc388.

> Why are we messing with people's environment at all? Shouldn't we respect people's .bashrc/.profile or whatever and not change colors and such?

Joe, we simply even didn't any use profiles. Thanks for pointing out, fixed in the PR here https://github.com/webmin/webmin/pull/1782.

**Reply by swelljoe** [reply]
Wait. That's not what I meant. When you start bash, it automatically loads these. We shouldn't be doing anything. We should let bash (or whatever the user's chosen shell is) do whatever it normally does.

**Reply by iliaross** [reply]
Something should source `~/.bash_profile` .. `~/.bash_profile` is automatically sourced only in login mode (like using `su - ` or `ssh`). It doesn't happen in our case. If you run `shopt login_shell` you will see that `bash` is not a login shell.

**Reply by swelljoe** [reply]
huh. That's weird. Maybe that's what we should be doing instead? Is there a reason we aren't?

**Reply by jcameron** [reply]
Maybe we need to set a flag to `bash` to make it a login shell? I seem to recall that's possible, but I can't remember what the correct flag is..

**Reply by swelljoe** [reply]
This seems helpful: https://askubuntu.com/questions/853395/where-is-the-login-shell-defined

And, there is a flag to bash, but I think we want all shells to be login shells, not just bash. I'm not entirely sure how that should be done.

**Reply by jcameron** [reply]
I think this is what I was looking for : https://superuser.com/questions/278859/dash-in-front-of-bash

**Reply by iliaross** [reply]
I'm confused now. I ran additional tests and it seems that `~/.bash_profile` is already being sourced with out the patch in PR.

What do you see on your end?

**Reply by iliaross** [reply]
.. in that case I will revert the initial patch to run `PS1` in env in shell server, and improve the logic for it in `index.cgi`, because login shell would always overwrite our flavors (i.e. `PS1`), and there is no way at that stage to control it correctly.

This is easily fixable although, I will make a change to this PR, to make sure that we never override user defined `PS1` (what Joe suggested initially) _but_ will use our `PS1` in case user didn't .. as vast majority of users won't set PS1 like never ever, so I want to make sure that it will be handled nicely out of the box, while respecting advanced user configurations as well.

**Reply by swelljoe** [reply]
We're not supposed to control it! We're supposed to respect a user's wishes.

**Reply by iliaross** [reply]
Actually, that's what I proposed and said! We will respect user's preferences but will suggest our colours in case it wasn't set at all, which also can be disabled from module config.

I don't want to have colorless ugly root prompt shown by default.

Anyone ever working with latest builds of Kali can clearly appreciate the difference, in regards of what Kali provides for default shell prompt and what other distros don't do.

**Reply by jcameron** [reply]
If we just want to use our PS1 if not set, isn't it simpler to do that in `shellserver.pl` rather than via command injection?

Also, what does this PS1 setting change exactly?

**Reply by iliaross** [reply]
> If we just want to use our PS1 if not set, isn't it simpler to do that in `shellserver.pl` rather than via command injection?

If that's possible as our `PS1` is getting overwritten when `~/.bashrc` is called.

> Also, what does this PS1 setting change exactly?

Primary bash prompt.

**Reply by iliaross** [reply]
> If that's possible as our `PS1` is getting overwritten when `~/.bashrc` is called.

Actually, it must be possible as we have access to `$shellfh` right there. Now we need to read `PS1` first from it, test and send our `PS1` ..

**Reply by iliaross** [reply]
Check [this PR](https://github.com/webmin/webmin/pull/1784) out, it will support currently set `PS1`.

**Reply by jcameron** [reply]
We really shouldn't change PS1 if the user explicitly sets it in .bashrc though. 

Explain to me again why we need to change PS1? Is it just for colors? If so, can't we do that in xterm.js ?

**Reply by iliaross** [reply]
> We really shouldn't change PS1 if the user explicitly sets it in .bashrc though.

Jamie [this PR](https://github.com/webmin/webmin/pull/1784) explicitly and solely does that. If use has set `PS1` (not using default) or has it unset we set ours (in `bash` only).

> Explain to me again why we need to change PS1? Is it just for colors?

Yes, for colors only. To make it stand out and more enjoyable to use.

> If so, can't we do that in xterm.js

I don't think it's possible, and again we cannot disrespect user's preferences.

**Reply by iliaross** [reply]
For the reference, [this PR](https://github.com/webmin/webmin/pull/1785) considers all of your suggestions.

**Reply by jcameron** [reply]
Colors are nice, but I feel like we're doing a lot of complex and fragile stuff to make it work! I'm honestly not sure it's worth the added complexity for a nice-to-have feature..

**Reply by iliaross** [reply]
I'd be very sad without colors. If you think the stuff in the PR above is really fragile and you cannot recommend to fix it, then we can set the default to No and just let people activate it manually.

Please comment on my PR. We'll take it from there..

**Reply by iliaross** [reply]
> I feel like we're doing a lot of complex and fragile stuff to make it work!

What about this one - https://github.com/webmin/webmin/pull/1786.

**Reply by jcameron** [reply]
Have a look at my two latest commits - I think that's the best way to do this.

**Reply by iliaross** [reply]
Jamie, wait. Could you unroll it? Because I have already been doing it the familiar way! https://github.com/webmin/webmin/pull/1786

**Reply by iliaross** [reply]
> Jamie, wait. Could you unroll it?

Conflicts resolved. No additional actions needed. Just review [the PR](https://github.com/webmin/webmin/pull/1786) and merge it please.

**Reply by jcameron** [reply]
Ok, I've merged the PR .. so we're good now!

**Reply by iliaross** [reply]
Perfect!
